synapse/storage/databases/main/event_push_actions.py

```diff
@@ -309,7 +309,45 @@ def add_thread_id_txn(
 
             # Update progress.
             processed_rows = txn.rowcount
-            progress[f"max_{table_name}_stream_ordering"] = max_stream_ordering
+            progress["max_event_push_actions_stream_ordering"] = max_stream_ordering
+            self.db_pool.updates._background_update_progress_txn(
+                txn, "event_push_backfill_thread_id", progress
+            )
+
+            return processed_rows
+
+        def add_thread_id_summary_txn(txn: LoggingTransaction) -> int:
+            min_user_id = progress.get("max_summary_user_id", "")
+            min_room_id = progress.get("max_summary_room_id", "")
+
+            sql = """
+            SELECT user_id, room_id FROM event_push_summary
+            WHERE (user_id, room_id) > (?, ?)
+                AND thread_id IS NULL
+            ORDER BY user_id, room_id
+            LIMIT 1
+            OFFSET ?
+            """
+
+            txn.execute(sql, (min_user_id, min_room_id, batch_size))
+            row = txn.fetchone()
+            if not row:
+                return 0
+
+            max_user_id, max_room_id = row
+
+            sql = """
+            UPDATE event_push_summary
+            SET thread_id = 'main'
+            WHERE
+                (?, ?) < (user_id, room_id) AND (user_id, room_id) < (?, ?)
```

Comment:
two exclusive bounds here. I think at least one of them should be inclusive.

...Looking at it, I think you need the upper bound to be inclusive? As when we run out of rows, we `return 0` so you need to have processed the upper item

Reply:
fuck, yes thanks